Forbid dashes in hostnames and /etc/fstab parsing improvements

keszybz opened the pull request.

poettering commented on src/resolve/test-resolved-etc-hosts.c:
```diff
@@ -19,38 +21,32 @@ static void test_parse_etc_hosts_system(void) {
+        fputs("::0 some.where some.other\n", f);
+        fputs("0.0.0.0 black.listed\n", f);
+        fputs("::5 some.where some.other foobar.foo.foo\n", f);
+        fputs("        \n", f);
```
i would probably have place this in a single string, rather than so many separate fputs() calls

keszybz replied:
This was already done in one of the later patches. I moved it earlier now.

poettering commented on src/resolve/test-resolved-etc-hosts.c:
```diff
@@ -19,38 +21,32 @@ static void test_parse_etc_hosts_system(void) {
+        fputs("0.0.0.0 black.listed\n", f);
+        fputs("::5 some.where some.other foobar.foo.foo\n", f);
+        fputs("        \n", f);
+        fflush(f);
```
maybe assert_se(fflush_and_check(f) == 0);

poettering commented on src/resolve/test-resolved-etc-hosts.c:
```diff
@@ -62,6 +65,17 @@ static void test_parse_etc_hosts(void) {
+
+        assert_se(bn->addresses[0]->family == AF_INET);
+        assert_se(memcmp(&bn->addresses[0]->address.in,
+                         &(struct in_addr) { .s_addr = htobe32(0x01020306) }, 4) == 0);
```
a slightly sloppy but more easily readable syntax for this could be: `._saddr = inet_addr("1.2.3.6")`...

might be good enough for a test case...

poettering commented on src/resolve/test-resolved-etc-hosts.c:
```diff
@@ -37,6 +37,9 @@ static void test_parse_etc_hosts(void) {
         assert_se(f);
         fputs("1.2.3.4 some.where\n", f);
         fputs("1.2.3.5 some.where\n", f);
+        fputs("1.2.3.6 dash dash-dash.where-dash\n", f);
```
hmm, why this patch? are you sure hostnames, with leading/trailing dashes are not OK? note that in dnssd and SRV RRs domain names with leading underscore are explicitly OK... 

What's you reasoning for this, is there any RFC or otherwise recommendation to disallow these?

grawity replied:
Dashes and underscores are ok in general, but if the entry has A/AAAA records (i.e. it's considered to be a "hostname") then many implementations – including Bind9 as reference – follow the more restrictive syntax from [RFC 1035 section 2.3.1](https://tools.ietf.org/html/rfc1035#section-2.3.1) which disallows leading/trailing dashes or underscores. For example, here's what `named-checkzone` says:

```
./test.zone:32: -ahost.example.com: bad owner name (check-names)
./test.zone:33: ahost-.example.com: bad owner name (check-names)
./test.zone:34: _ahost.example.com: bad owner name (check-names)
./test.zone:35: ahost_.example.com: bad owner name (check-names)
./test.zone:36: a_host.example.com: bad owner name (check-names)
```

Whereas if the entry only has "non-host" records (like SRV or TXT or PTR), then everything is allowed.

keszybz commented:
Updated, with all suggestion implemented.

For the "no dashes" check: I now restricted it to /etc/hosts only. Here we should generally follow what nss-files does, and both the docs and actual nss-files implementation agree that dashes are not allowed. Longer term, we could apply the filtering also to TLDs and A/AAAA records.

poettering commented on src/shared/dns-domain.c:
```diff
@@ -52,6 +54,8 @@ int dns_label_unescape(const char **name, char *dest, size_t sz) {
                 if (*n == '\\') {
                         /* Escaped character */
+                        if (FLAGS_SET(flags, DNS_LABEL_LDH))
+                                return -EINVAL;
```
this sounds overly strict, no? it sounds OK to allow escaped chars here, no?

poettering commented on src/shared/dns-domain.c:
```diff
@@ -36,13 +36,15 @@ int dns_label_unescape(const char **name, char *dest, size_t sz) {
-                        break;
-                }
+                if (*n == 0 || *n == '.') {
+                        if (FLAGS_SET(flags, DNS_LABEL_LDH) && r > 0 && n[-1] == '-')
```
hmm, it feels a bit sloppy to check the last character in the *source* string, i.e. in n[-1], because we apply unescaping on that after all... (well, we don't allow escapes with your patch, but I think we should). I'd prefer if we'd just have a new local variable `last_char` or so that always contains the last char that was added to the output string, that we can check here.

poettering commented on src/shared/dns-domain.c:
```diff
@@ -103,6 +107,10 @@ int dns_label_unescape(const char **name, char *dest, size_t sz) {
+                        if (FLAGS_SET(flags, DNS_LABEL_LDH) && r == 0 && *n == '-')
+                                /* Leading dash */
+                                return -EINVAL;
+
```
Hmm, this current function is happy to accept non-7bit chars too (it has too, mdns needs that). I figure when the flag is set we should refuse that too, in fact following the excerpt of the RFC you quoted we should probably refuse a lot more even, including non-alphanumerics and stuff

poettering replied:
(that said, does glibc honour escapes in /etc/hosts? if not, then we probably should not honour these either here. But I think that should be a separate flag then, i.e. DNS_LABEL_NO_ESCAPES or so, that the /etc/hosts parser then sets.

keszybz replied:
It doesn't:
```console
$ grep abc /etc/hosts                                                                                            
1.2.4.5  a\097a abc                                                                                              
$ getent hosts aaa                                                                                               
$ getent hosts $'a\\097a'                                                                                        
1.2.4.5         a\097a abc                                                                                       
```

poettering commented:
btw, i figure we should enforce the LDH rule in hostname-util.c's hostname_is_valid() too. After all that function is supposed to validate hostnames by the strictest possible rules, include Linux-specific hostname rules (such as the size limit).

keszybz commented:
Updated. I split the flag in two as requested: `DNS_LABEL_LDH` and `DNS_LABEL_NO_ESCAPES`. I also made `hostname-util.c` use the same rules. I didn't fully unify the implementations, because `hostname-util.c` has both `hostname_is_valid()` and `hostname_cleanup()`, and the additional length and dot requirements. Porting `hostname_is_valid()` to use `dns_name_valid()` wouldn't be easy and anyway I think it's better to keep `hostname_is_valid()` and `hostname_cleanup()` aligned. I also added more tests.

poettering commented on src/resolve/test-resolved-etc-hosts.c:
```diff
@@ -21,6 +27,14 @@ static void test_parse_etc_hosts_system(void) {
+
+#define address_equal_6(_addr, ...)                           \
+        ((_addr)->family == AF_INET6 &&                                 \
+         !memcmp(&(_addr)->address.in6, &(struct in6_addr) { .s6_addr = __VA_ARGS__}, 16) )
```
double evaluation sucks...

poettering replied:
(but given this is just a test doesn't really matter)

keszybz replied:
Yeah, in this particular case I think this simple form is enough.

poettering commented on src/resolve/test-resolved-etc-hosts.c:
```diff
@@ -21,6 +27,14 @@ static void test_parse_etc_hosts_system(void) {
+        ((_addr)->family == AF_INET &&                                  \
+         !memcmp(&(_addr)->address.in, &(struct in_addr) { .s_addr = (_address) }, 4))
+
+#define address_equal_6(_addr, ...)                           \
```
whitespace is off here

poettering replied:
(but doesn't really matter)

poettering commented on src/resolve/resolved-etc-hosts.c:
```diff
@@ -164,7 +165,7 @@ int etc_hosts_parse(EtcHosts *hosts, FILE *f) {
         int r;
 
-        for (;;) {
+        for (nr = 0; ; nr++) {
```
hmm, this means line numbering now starts at zero, right? this is probably not what most people expect. Also nr is assigned zero a few lines up already, one should probably go

poettering commented:
so the line nr thing matters i guess, the other two nitpicks i found don't

keszybz commented:
Updated to fix the whitespace issue and off-by-one in line counting.

poettering commented:
lgtm

poettering commented:
So, this caused #11765 btw...
